### ISSUE TITLE
Uri Fragment Route Giving Params Error

### ISSUE THREAD

**Issue by LaserStony** [apioo/fusio#168]
I created an Action to get all the Projects. Then I created an Action for the Route. But when I execute the UriFragment Action, I get the Following Error.:

`"An exception occurred while executing 'SELECT * FROM app_todo WHERE AppID = :appid' with params [\"12\"]:\n\nValue for :appid not found in params array. Params array key should be \"appid\" in /home/website_name/public_html/api/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php on line 131"`

So, Do I need to create some Params?

**Comment by chriskapp** [comment]
Yes, so if you execute the query you need to provide an array containing the parameters i.e. `['appId' => 12]`. Otherwise please show me the code of your action then I could help to debug the problem.

**Comment by LaserStony** [comment]
Got It. Thanks! Was Passing `id `instead of `appId `in the Parameters.

**Comment by LaserStony** [comment]
@chriskapp, How can I use the Authorization System? I Have a Route i.e: `http://api.mywebsite.com/public/index.php/Projects/projectname`
which works perfectly when I use it publicly. But when I un-check the Public checkbox from its settings, I get Errors. for instance `Access to this resource is not in the scope of the provided token in /home/website/public_html/api/vendor/fusio/impl/src/Filter/Authentication.php on line 200`
The Token is passed Correctly. I checked. The User has all three Scopes Assigned to it. Still, I cannot seem to Get it to work..

**Comment by chriskapp** [comment]
Hi, so if you mark your route as private you need to pass a fitting access token as header in the request i.e. `Authorization: Bearer [access token]`. There are multiple ways to obtain such an access token. The simplest way is to call the `/consumer/login`endpoint like described in this [doc](https://www.fusio-project.org/bootstrap). Its important that you user account has assigned the fitting scopes and that also the consumer app has the scopes selected. Then you should be able to access the protected endpoint.

Regarding associations, yes you simple need to write the fitting SQL query. If you show me the structure of the tables I could help you to find the right query but it mostly looks like:

```
    SELECT * 
      FROM Product 
INNER JOIN Variations
        ON Variations.Id = Product.VariationsId 
INNER JOIN AvailabilityStatus
        ON AvailabilityStatus.Id = Product.AvailabilityStatusID 
     WHERE Product.Id = :id
```

**Comment by LaserStony** [comment]
I get the token from the consumer/login by sending the user as the body (on postman) and then I pass the tomen as header to the above mentioned request, but still get the error I mentioned above. I want to know if there is a different way for accessing private route instead of `public` or is it the only way. Am I doing something wrong? I followed the Doc before posting this. And I made sure to follow it to the best of my capabilities.

**Comment by chriskapp** [comment]
> I get the token from the consumer/login by sending the user as the body (on postman) and then I pass the token as header to the above mentioned request, but still get the error I mentioned above. Am I doing something wrong?

Mhh ok, so to debug this problem further: Do you have assigned the route to a scope? Is this scope also assigned to your user and the consumer app? You could also take a look at the `fusio_app_token` table to see which scopes are assigned to the token. If this is all fine you should be able to request the protected endpoint with your access token.

> Also, by using the above Query I get the data in a single Json Object. While I want to get multiple Json Objects.

Since it is very common to build nested structures based on flat tables, Fusio has a dedicated builder class which helps to solve exactly this problem. In the following a simple script which shows how this could work:

```php
/** @var \Doctrine\DBAL\Connection $connection */
$connection = $connector->getConnection('My-Connection');

$builder = new \PSX\Sql\Builder($connection);

$definition = $builder->doEntity('SELECT * FROM Product WHERE Product.Id = :id', ['id' => $request->getUriFragment('id')], [
    'id' => $builder->fieldInteger('id'),
    'Name' => 'Name',
    'Type' => 'Type',
    'Variations' => $builder->doCollection('SELECT * FROM Variations WHERE id = :id', ['id' => new \PSX\Sql\Reference('VariationsId')], [
        'id' => $builder->fieldInteger('id'),
        'Name' => 'Name',
    ]),
    'AvailabilityStatus' => $builder->doEntity('SELECT * FROM AvailabilityStatus WHERE id = :id', ['id' => new \PSX\Sql\Reference('AvailabilityStatusID')], [
        'id' => $builder->fieldInteger('id'),
        'Name' => 'Name',
        'Abbr' => 'Abbr',
        'Desc' => 'Desc',
    ],
]);

return $response->build(200, [], $builder->build($definition));
```

Please note I hove wrote this just from the top of my head so it might contain errors.

**Comment by LaserStony** [comment]
> Mhh ok, so to debug this problem further: Do you have assigned the route to a scope? Is this scope also assigned to your user and the consumer app? You could also take a look at the fusio_app_token table to see which scopes are assigned to the token. If this is all fine you should be able to request the protected endpoint with your access token.

Got this over with. I think it was the Assigning of Scope to the Route...

Now will test the Json Builder one in the Morning and will report on it as well... Thanks Again!

**Comment by LaserStony** [comment]
@chriskapp, I cannot seem to get it to work.. It Gives me `Reference invalid context key ` and the the Key name. It works fine for the first key but not for the ones with Null Values. Is it possible for it to return null/empty object if the Value is null or it doesn't Find a Reference?

**Comment by chriskapp** [comment]
ok, if you get this error this means that the key which you have provided in a reference i.e. `\PSX\Sql\Reference('VariationsId')` is not available in the parent context. In this case the column `VariationsId` must exists as column of the `SELECT * FROM Product` query. So I think that a specific column in such a reference does not exist.

Regarding the null values, Iam not sure whether I understand the problem correctly. Could you show me the code, then it is probably easier for me to understand.

**Comment by LaserStony** [comment]
@chriskapp I went through the syntax quiet Briefly. And Reached a Conclusion. Now see:

```
/** @var \Doctrine\DBAL\Connection $connection */
$connection = $connector->getConnection('My-Connection');

$builder = new \PSX\Sql\Builder($connection);

$definition = $builder->doEntity('SELECT * FROM Product WHERE Product.Id = :id', ['id' => $request->getUriFragment('id')], [
    'id' => $builder->fieldInteger('id'),
    'Name' => 'Name',
    'Type' => 'Type',
    'Variations' => $builder->doCollection('SELECT * FROM Variations WHERE id = :id', ['id' => new \PSX\Sql\Reference('VariationsId')], [
        'id' => $builder->fieldInteger('id'),
        'Name' => 'Name',
    ]),
    'AvailabilityStatus' => $builder->doEntity('SELECT * FROM AvailabilityStatus WHERE id = :id', ['id' => new \PSX\Sql\Reference('AvailabilityStatusID')], [
        'id' => $builder->fieldInteger('id'),
        'Name' => 'Name',
        'Abbr' => 'Abbr',
        'Desc' => 'Desc',
    ],
]);

return $response->build(200, [], $builder->build($definition));
```
Now, I get `"Referenced unknown key \"Abbr\" in context` If it's Value is `null `and in **DB** it is set to be `not-Null`. Same for `VariationsId `as Well. Once I change the values to something rather than Null, it works fine and shows the Result. 
Also, If the value of a field is `Null` and in **DB** it is set to be `Null` then that field is completely skipped by the Builder.
So, Is it possible for someway for it to show the values if the value is null?

**Comment by LaserStony** [comment]
@chriskapp can you give me some pointers in the right direction? I am unable to continue because of this. thanks!

**Comment by chriskapp** [comment]
Hi @LaserStony for sure, so I understand your problem. Currently the builder cant handle null values this way it expects that the values must be i.e. a string, int, etc. since it uses `isset` to check whether a property exists. Which does not work for `null` values. As a workaround you could use something like `IFNULL(Abbr, '')` in your SQL query to convert null to an empty string. If you like you can also try to update the `psx/sql` component to the latest version with `composer update psx/sql` which improves the `null` handling a bit.

**Comment by LaserStony** [comment]
H!... Is there any way to update psx/sql Manually? cause I don't Have the necessary control over the server where I have installed it. I can use FTP though. So, will it work if I uploaded the new version to vendor/psx in my Fusio Installation on the Server?

**Comment by chriskapp** [comment]
Yes, of course, you can also download the latest release from the repo https://github.com/apioo/psx-sql and replace it at the vendor folder under `vendor/psx/sql` this has basically the same effect as composer update. In fact since your are using FTP it would be enough to only update only the `src/Builder.php` file.

**Comment by LaserStony** [comment]
@chriskapp Thanks Man! I will go and do it now. Thanks Again for helping me out! (Hoping that this would solve all problems)

**Comment by LaserStony** [comment]
@chriskapp, Nice... It doesn't Give the Null error.. but Still, It removes the Field. Wouldn't It give me an error if I used it With result without that Query in my Application for Parsing?
